### PR TITLE
Conda pack support

### PR DESCRIPTION
**Pull request by PiotrZierhoffer**
This PR is dependent on https://github.com/timvideos/litex-buildenv/pull/212

It adds two scripts: build-pack.sh and enter-pack.sh that allow to create and use the package.

build-pack.sh can be started on a clean lxbe repo, building a package in ./build/lxbe-env.tar.gz. It weights about 1.3GB.

The pack contains a Conda environment, which is subsequently used by enter-pack.

enter-pack.sh must be sourced and requires the usual variables to be set (e.g. `export TARGET=base FIRMWARE=zephyr PLATFORM=arty CPU_VARIANT=full CPU=vexriscv`). It unpacks the provided tar.gz (or uses the already unpacked directory) and activates the environment.

This is an important change to the way enter-env.sh operates (that is - using the default env, without activation).

After the activation, the usual tools can be used to build/run/flash things.

### CODE REVIEW

File: scripts/template-environment.yml [hunk 1]
```diff
@@ -0,0 +1,18 @@
+name: lxbe-env
+channels:
+  - timvideos
+dependencies:
+  - fxload
+  - icefunprog
+  - iceprog
+  - ninja
```
[the remark this answers is not in the document]
Comment (mateusz-holenko) [reply scripts/template-environment.yml]: Both `gperf` and `ninja` are needed for Zephyr SDK

File: scripts/template-environment.yml [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+  - gperf
```
[the remark this answers is not in the document]
Comment (mateusz-holenko) [reply scripts/template-environment.yml]: Both `gperf` and `ninja` are needed for Zephyr SDK

[the remark this answers is not in the document]
Comment (mithro) [reply scripts/template-environment.yml]: Can you add some comments?

File: scripts/build-pack.sh [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+			west \
+			pykwalify \
+			"git+https://github.com/tinyfpga/TinyFPGA-Bootloader#egg=tinyprog&subdirectory=programmer&subdirectory=programmer" \
+			git+https://github.com/mithro/hexfile.git \
+			git+https://github.com/timvideos/HDMI2USB-mode-switch.git \
+			"cmake==$CMAKE_VERSION"
+
+
+# This has to be installed manually in the environment
+MIMASV2CONFIG=$BUILD_DIR/conda/bin/MimasV2Config.py
```
[the remark this answers is not in the document]
Comment (PiotrZierhoffer) [reply scripts/build-pack.sh]: As I don't think it would make sense to PR to the original repo, do you want to have it in timvideos GH?

File: scripts/build-pack.sh [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+cat $ENV_TEMPLATE > $ENV_FILE
+echo "  - conda=$CONDA_VERSION" >> $ENV_FILE
+echo "  - python=$PYTHON_VERSION" >> $ENV_FILE
+echo "  - openocd=$OPENOCD_VERSION" >> $ENV_FILE
+
+# This code is supposed to add entries for all architectures and toolchain
+# flavors.
+# Some conda packages are not there yet (namely x86_64 and lm32 with musl),
+# so we comment these out.
+for ARCH in or1k riscv32 lm32 x86_64; do
```
[the remark this answers is not in the document]
Comment (PiotrZierhoffer) [reply scripts/build-pack.sh]: Not per se, it was actually an emulation of `CPU=none`. Is there any reasonable usage of this setting at this moment?

File: scripts/build-pack.sh [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+# match each other
+ENV_NAME=lxbe-env
+
+# As conda environment yml format does not seem to allow variables and we want
+# to rely on settings.sh to specify package versions, we programatically fill up
+# the env file with proper versions.
+cat $ENV_TEMPLATE > $ENV_FILE
+echo "  - conda=$CONDA_VERSION" >> $ENV_FILE
+echo "  - python=$PYTHON_VERSION" >> $ENV_FILE
+echo "  - openocd=$OPENOCD_VERSION" >> $ENV_FILE
```
[the remark this answers is not in the document]
Comment (PiotrZierhoffer) [reply scripts/build-pack.sh]: I think it makes sense, parsing it will be trivial and will allow us to drop this quasi-auto-generation of yml. It's a shame conda does not allow variables substitution. Even python-pip does.

File: scripts/build-pack.sh [hunk 6]
```diff
@@ -0,0 +1,155 @@
+#!/bin/bash
+
+### This script is, in principle, similar to download-env.sh
+### Notable differences:
+### - download-env does not download everything, but depends on system variables
+### - download-env uses user's site-packages
```
[the remark this answers is not in the document]
Comment (PiotrZierhoffer) [reply scripts/build-pack.sh]: I will double check it, but I believe I was able to use a local package as a dependency.

**Comment by mithro** [comment]
Please rebase onto master.

File: scripts/template-environment.yml [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+  - pyyaml
+  - ipython
+  - pyserial
```
[the remark this answers is not in the document]
Comment (PiotrZierhoffer) [reply scripts/template-environment.yml]: template-environment.yml is used to create a one-pack-that-has-all. So the idea here was to gather all the deps in one place. Is this in line with what you wanted to have?

File: scripts/template-environment.yml [hunk 8]
```diff
@@ -0,0 +1,18 @@
+name: lxbe-env
+channels:
+  - timvideos
+dependencies:
+  - fxload
+  - icefunprog
+  - iceprog
```
[the remark this answers is not in the document]
Comment (PiotrZierhoffer) [reply scripts/template-environment.yml]: see above

File: scripts/build-pack.sh (hunk 5 above)
[the remark this answers is not in the document]
Comment (PiotrZierhoffer) [reply scripts/build-pack.sh]: @mithro So we gave some thought to it and this calls for more discussion.

- download-env.sh installs packages on per-need basis
- build-pack.sh installs all the packages at once
- the versions are the same in both scenarios
- some packages are downloaded via pip, some are via conda.

Right now cmake is the only pip-installed package that has a version specified. It turns out that Conda has it in an older, but still suitable version available.

This means we could have multiple requirements.txt files in scripts/dependencies. There could be requirements.txt for common stuff, requirements-zephyr.txt, requirements-tinyfpga_b2.txt etc.

This way we could:
a) have a clear distinction over what requires which packages
b) install via ``pip install -r`` in both download-env and build-pack.

For Conda we can do it similarly, and have multiple requirement files for different dependencies, and then use them all to create a pack.
There are, allegedly, some problems with using files as source for ``conda install``, but I'm sure it's manageable. The format is also not really specified, but I think that ``pip`` files work fine, just as yml files.

Parsing files is something we'd prefer to avoid, for example to be able to easily switch the format/backend in the future.

Summing up, the flow would be:

```
download-env.sh:
if zephyr then ensure_deps_for(zephyr)
if sth_else then ensure_deps_for(sth_else)

build-pack.sh:
ensure_all_deps()

ensure_deps_for(x):
    // in future: support docker/local installations
    conda install -f conda_deps_for_$x
    pip install -r pip_deps_for_$x

ensure_all_deps():
   for x in deps:
      ensure_deps_for(x)
```

What's your opinion on that?

File: scripts/build-pack.sh [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+
+set -u
+set -e
+
+# We're using pip install instead of putting it in the yml, because we want to
+# ensure isolation.
+# conda env create does not seem to take PYTHONNOUSERSITE=1 very seriously.
+# We're not installing conda-pack from conda because it requires conda-forge
+# source.
+pip install conda-pack \
```
[the remark this answers is not in the document]
Comment (PiotrZierhoffer) [reply scripts/build-pack.sh]: please see https://github.com/timvideos/litex-buildenv/pull/213#discussion_r343540545

File: scripts/build-pack.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (mithro) [reply scripts/build-pack.sh]: This should not be possible.

File: scripts/build-pack.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (mithro) [reply scripts/build-pack.sh]: It does make sense to PR to the original repo, they will probably accept the pull request.

I have contacts at Numato if things go south.

File: scripts/enter-pack.sh [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+export PYTHONNOUSERSITE=1
+
+echo ""
+echo "Activating environment"
+echo "----------------------"
+
+source $ENV_PATH/bin/activate
+conda-unpack
+
+# Some packages we use (like openocd or iceprog) rely on host GCC and binutils
```
Comment (mithro) [inline scripts/enter-pack.sh]: Can we fix them so they don't rely on host gcc / binutils?

File: scripts/build-pack.sh (hunk 5 above)
[the remark this answers is not in the document]
Comment (mithro) [reply scripts/build-pack.sh]: I like the idea of having;
 * `requirements-<board>.txt`
 * `requirements-<firmware>.txt`
 * `requirements-<cpu>.txt`

I'm unclear if it should be separate firmware / cpu files or if it should be `requirements-<firmware>-<cpu>`?

Another option is that we could use dummy conda packages with the requirements as dependencies?

[the remark this answers is not in the document]
Comment (PiotrZierhoffer) [reply scripts/build-pack.sh]: I believe I will be enough to have `requirements-<something>.txt`, with only one dimension. If more is needed, we can always add it.

File: scripts/build-pack.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (PiotrZierhoffer) [reply scripts/build-pack.sh]: I think that PYTHONNOUSERSITE=1 you added in eb927b8 fixes it

File: scripts/enter-pack.sh (hunk 10 above)
Comment (PiotrZierhoffer) [reply scripts/enter-pack.sh]: If I understand correctly, the reason for these dependencies is here: https://github.com/litex-hub/litex-conda-packages/blob/master/lib/ftdi/meta.yaml#L31-L33

I can try removing these dependencies unless you see an obvious reason not to

File: scripts/build-pack.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (PiotrZierhoffer) [reply scripts/build-pack.sh]: How can it be used in a meaningful way? I mean, how do we define "works"?

[the remark this answers is not in the document]
Comment (mithro) [reply scripts/build-pack.sh]: For a compiler, test it is able to compile something?

[the remark this answers is not in the document]
Comment (PiotrZierhoffer) [reply scripts/build-pack.sh]: Do you mean x86 compiler? I don't really understand the way you'd like to use it. `CPU=none ./scripts/build-linux.sh` could make sense in a way, but it has nothing to do with LiteX or the compilers we usually use, it's just a generic kernel compilation within a conda env. The same applies to Zephyr or micropython.

To effectively include this option in the general flow I need to understand how can it be useful.

**Comment by mithro** [comment]
@mateusz-holenko -- Could you work with @PiotrZierhoffer to get something like this pull request merged?
